apps/rush-lib/src/api/RushConfigurationProject.ts

```diff
@@ -227,25 +228,55 @@ export class RushConfigurationProject {
 
   /**
    * A list of projects within the Rush configuration which directly depend on this package.
+   * @deprecated Use downstreamDependencyProjectSet instead
    */
   public get downstreamDependencyProjects(): string[] {
+    return [...this._downstreamDependencyProjects];
+  }
+
+  /**
+   * A set of projects within the Rush configuration which directly depend on this package.
+   */
+  public get downstreamDependencyProjectSet(): Set<string> {
     return this._downstreamDependencyProjects;
   }
 
   /**
    * A map of projects within the Rush configuration which are directly depended on by this project
+   * @deprecated Use localDependencyProjectSet instead
    */
   public get localDependencyProjects(): ReadonlyArray<RushConfigurationProject> {
+    return [...this.localDependencyProjectSet];
+  }
+
+  /**
+   * The set of projects within the Rush configuration which are directly depended on by this project
+   */
+  public get localDependencyProjectSet(): ReadonlySet<RushConfigurationProject> {
     if (!this._localDependencyProjects) {
-      this._localDependencyProjects = [
-        ...this._getLocalDependencyProjects(this.packageJson.dependencies),
-        ...this._getLocalDependencyProjects(this.packageJson.devDependencies),
-        ...this._getLocalDependencyProjects(this.packageJson.optionalDependencies)
-      ];
+      const self: RushConfigurationProject = this;
+      this._localDependencyProjects = new Set(
+        (function* () {
+          yield* self._getLocalDependencyProjects(self.packageJson.dependencies);
+          yield* self._getLocalDependencyProjects(self.packageJson.devDependencies);
+          yield* self._getLocalDependencyProjects(self.packageJson.optionalDependencies);
+        })()
```

Comment:
Just concatenate the arrays with spread operators. This is too complicated.

Reply:
Generators being too complicated is a self-fulfilling prophecy. If no one ever uses them for things they do efficiently, no one will ever get used to them. Much like the "if you assume all your users can't understand URLs, they will never learn to understand URLs" issue of modern browsers.

Reply:
Switched to `Selection.union()` here because I already have a helper function that does this.